java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventData.java

```diff
@@ -196,43 +207,130 @@ public EventData(ByteBuffer buffer)
 	{
-		return this.properties;
-	}
+		if (this.properties == null)
+		{
+			this.properties = new HashMap<String, String>();
+		}
 
-	public void setProperties(Map<String, String> applicationProperties)
-	{
-		this.properties = applicationProperties;
+		return this.properties;
 	}
 
 	/**
 	 * SystemProperties that are populated by EventHubService.
 	 * <p>As these are populated by Service, they are only present on a Received EventData.
 	 * @return an encapsulation of all SystemProperties appended by EventHubs service into EventData
+	 * @deprecated As of release 0.8.0. Use {@link #getSystemProperties()}
 	 */
-	public SystemProperties getSystemProperties()
+	@Deprecated public SystemProperties getDeprecatedSystemProperties()
 	{
-		if (this.isReceivedEvent && this.systemProperties == null)
+		if (this.isReceivedEvent && this.deprecatedSystemProperties == null)
 		{
-			this.systemProperties = new SystemProperties(this);
+			this.deprecatedSystemProperties = new SystemProperties(this);
 		}
 
+		return this.deprecatedSystemProperties;
+	}
+	
+	/**
+	 * {@link #getSystemProperties()} from version 0.7.9 was deprecated and renamed to {@link #getDeprecatedSystemProperties()} starting 0.8.0.
+	 * {@link #getOffset()}, {@link #getSequenceNumber()}, {@link #getPartitionKey()}, {@link #getEnqueuedTime()} are now available directly on {@link EventData} class.
+	 * <p>SystemProperties that are populated by EventHubService.
+	 * <p>As these are populated by Service, they are only present on a Received EventData.
+	 * <p>Usage:<p>
+	 * <code>
+	 * final String offset = eventData.getSystemProperties().get(AmqpConstants.OFFSET_ANNOTATION_NAME);
+	 * </code>
+	 * @return an encapsulation of all SystemProperties appended by EventHubs service into EventData.
+	 * <code>null</code> if the {@link #EventData()} is not received and is created by the public constructors. 
+	 */
+	public Map<String, Object> getSystemProperties()
+	{
 		return this.systemProperties;
 	}
-
+	
+	public String getOffset()
+	{
+		return this.getSystemProperty(AmqpConstants.OFFSET_ANNOTATION_NAME);
+	}
+	
+	public String getPartitionKey()
+	{
+		return this.getSystemProperty(AmqpConstants.PARTITION_KEY_ANNOTATION_NAME);
+	}
+	
+	public Instant getEnqueuedTime()
+	{
+		final Date enqueuedTimeValue = this.getSystemProperty(AmqpConstants.ENQUEUED_TIME_UTC_ANNOTATION_NAME);
+		return enqueuedTimeValue != null ? enqueuedTimeValue.toInstant() : null;
+	}
+	
+	public long getSequenceNumber()
+	{
+		return this.getSystemProperty(AmqpConstants.SEQUENCE_NUMBER_ANNOTATION_NAME);
+	}
+	
+	@SuppressWarnings("unchecked")
+	private <T> T getSystemProperty(final String key)
+	{
+		if (this.systemProperties != null && this.systemProperties.containsKey(key))
+		{
+			return (T) (this.systemProperties.get(key));
+		}
+		
+		return null;
+	}
+	
+	// This is intended to be used while sending EventData - so EventData.SystemProperties will not be copied over to the AmqpMessage
```

Comment:
>EventData.SystemProperties will not be copied over to the AmqpMessage [](start = 61, length = 69)

but what is the harm of doing it anyway? I can think of things such as middle-tier gateway that would want everything even in send.